File: fithub-frontend/src/types/AssignedWorkoutPlan.ts
import { type WorkoutPlan } from "./WorkoutPlan";

export type AssignmentStatus = "ACTIVE" | "COMPLETED" | "CANCELLED";

export interface AssignedWorkoutPlan {
  id: number;
  memberId: number;
  workoutPlan: WorkoutPlan; // Nested full workout plan details
  assignedByTrainerId?: number;
  startDate: string; // ISO date string
  endDate?: string; // ISO date string, optional
  status: AssignmentStatus;
  createdAt: string;
  updatedAt: string;
}

export interface AssignedWorkoutPlanRequest {
  memberId: number;
  planId: number;
  assignedByTrainerId?: number;
  startDate: string; // ISO date string
  endDate?: string; // ISO date string
  status?: AssignmentStatus;
}

File: workout-service/src/main/java/com/gym/workoutservice/entity/AssignedWorkoutPlan.java
package com.gym.workoutservice.entity;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "assigned_workout_plan")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class AssignedWorkoutPlan {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "member_id", nullable = false)
    private Long memberId; // Member ID from User Service

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "plan_id", nullable = false)
    private WorkoutPlan workoutPlan;

    @Column(name = "assigned_by_trainer_id")
    private Long assignedByTrainerId; // Trainer ID from User Service

    @Column(name = "start_date", nullable = false)
    private LocalDate startDate;

    @Column(name = "end_date")
    private LocalDate endDate; // Optional, can be null for ongoing plans

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private AssignmentStatus status = AssignmentStatus.ACTIVE;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    public enum AssignmentStatus {
        ACTIVE, COMPLETED, CANCELLED
    }
}
